File: frontend/src/services/notification.service.ts
import api from '@/services/api';
import type { ApiResponse, NotificationListResponse, NotificationSummary } from '@/types';

export const notificationService = {
  async getNotifications(page = 1, limit = 10, unreadOnly = false): Promise<NotificationListResponse> {
    const params = new URLSearchParams({
      page: String(page),
      limit: String(limit),
    });
    if (unreadOnly) {
      params.set('unread', 'true');
    }

    const response = await api.get<ApiResponse<NotificationListResponse>>(
      `/notifications?${params.toString()}`
    );
    return response.data.data ?? {
      notifications: [],
      total: 0,
      page,
      page_size: limit,
      total_pages: 0,
    };
  },

  async getSummary(): Promise<NotificationSummary> {
    const response = await api.get<ApiResponse<NotificationSummary>>('/notifications/summary');
    return response.data.data ?? { unread_count: 0 };
  },

  async markRead(notificationId: string): Promise<void> {
    await api.put(`/notifications/${notificationId}/read`, {});
  },

  async markAllRead(): Promise<void> {
    await api.put('/notifications/read-all', {});
  },
};

File: backend/internal/services/notification/notification_service.go
package notification

import (
	"ecommerce-backend/internal/models"
	"ecommerce-backend/internal/repositories"
	"encoding/json"

	"github.com/google/uuid"
)

type Service struct {
	repo *repositories.NotificationRepository
}

func NewService(repo *repositories.NotificationRepository) *Service {
	return &Service{repo: repo}
}

func (s *Service) CreateForUser(userID uuid.UUID, notificationType string, title string, message string, metadata map[string]interface{}) error {
	rawMetadata := "{}"
	if metadata != nil {
		encoded, err := json.Marshal(metadata)
		if err != nil {
			return err
		}
		rawMetadata = string(encoded)
	}

	return s.repo.Create(&models.Notification{
		ID:       uuid.New(),
		UserID:   userID,
		Type:     notificationType,
		Title:    title,
		Message:  message,
		Metadata: rawMetadata,
	})
}

func (s *Service) ListForUser(userID uuid.UUID, page, pageSize int, unreadOnly bool) (*models.NotificationListResponse, error) {
	notifications, total, err := s.repo.ListByUser(userID, page, pageSize, unreadOnly)
	if err != nil {
		return nil, err
	}
	totalPages := (int(total) + pageSize - 1) / pageSize

	return &models.NotificationListResponse{
		Notifications: notifications,
		Total:         int(total),
		Page:          page,
		PageSize:      pageSize,
		TotalPages:    totalPages,
	}, nil
}

func (s *Service) GetSummary(userID uuid.UUID) (*models.NotificationSummaryResponse, error) {
	total, err := s.repo.CountUnread(userID)
	if err != nil {
		return nil, err
	}
	return &models.NotificationSummaryResponse{UnreadCount: int(total)}, nil
}

func (s *Service) MarkRead(userID uuid.UUID, notificationID uuid.UUID) error {
	return s.repo.MarkRead(userID, notificationID)
}

func (s *Service) MarkAllRead(userID uuid.UUID) error {
	return s.repo.MarkAllRead(userID)
}
